### DISCUSSION TITLE
Partitioned solver results

### DISCUSSION THREAD

**Discussion by paraynaud** [paraynaud/PartiallySeparableSolvers.jl#20] [Show and tell]
@dpo I made a script to gather the unconstrained partially separable problems from the current version of OptimizationProblems.

```julia
using PartiallySeparableSolvers
using JSOSolvers, SolverBenchmark, SolverTools, Plots, DataFrames, NLPModels
using CSV

n = 10
using ADNLPModels
using OptimizationProblems, OptimizationProblems.ADNLPProblems
using PartiallySeparableSolvers

problems_names = OptimizationProblems.meta[!,:name]
problems_symbols = map(name -> Symbol(name), problems_names)
problems = map( symbol -> getfield(OptimizationProblems.ADNLPProblems, symbol)(), problems_symbols)

res_isps = []
index_is_ps = []
for (index,nlp) in enumerate(problems)
  println(index)
  if (nlp.meta.ncon == 0) && (nlp.meta.name != "brownden")
    try
      (ex, n, x0) = PartiallySeparableSolvers.get_expr_tree(nlp)
      part_data_pqn = PartiallySeparableSolvers.build_PartitionedDataTRPQN(ex, n)
      bool = (part_data_pqn.n != 1) && (part_data_pqn.N != 1 )
      push!(res_isps, bool)
      push!(index_is_ps, index)
    catch 
      println("the ", index, "-th problem is unsupported")
    end
  end
end 
list_ps_supported = filter(i-> res_isps[i], 1:length(index_is_ps))
list_isps = index_is_ps[list_ps_supported]
name_ps= problems_names[list_isps]

ps_problems = (eval(Meta.parse(problem))(; n) for problem ∈ name_ps)
```
Ending up with 85 problems.
I run `bmark_solver` on them and compare `trunk` with partitioned methods, including the PCS method with:
```julia
solver = Dict{Symbol,Function}(

solver = Dict{Symbol,Function}(
  :PBFGS => ((prob;kwargs...) -> PartiallySeparableSolvers.PTRUNK(prob; name=:pbfgs, verbose=0, kwargs...)),
  :PLBFGS => ((prob;kwargs...) -> PartiallySeparableSolvers.PTRUNK(prob; name=:plbfgs, verbose=0, kwargs...)),
  :PSR1 => ((prob;kwargs...) -> PartiallySeparableSolvers.PTRUNK(prob; name=:psr1, verbose=0, kwargs...)),
  :PLSR1 => ((prob;kwargs...) -> PartiallySeparableSolvers.PTRUNK(prob; name=:plsr1, verbose=0, kwargs...)),
  :PSE => ((prob;kwargs...) -> PartiallySeparableSolvers.PTRUNK(prob; name=:pse, verbose=0, kwargs...)),
  :PLSE => ((prob;kwargs...) -> PartiallySeparableSolvers.PTRUNK(prob; name=:plse, verbose=0, kwargs...)),
  :PCS => ((prob;kwargs...) -> PartiallySeparableSolvers.PTRUNK(prob; name=:pcs, verbose=0, kwargs...)),
  :TRUNK => ((prob;kwargs...) -> JSOSolvers.trunk(prob; kwargs...)),
)

const max_time = 10.0
const max_eval = 10000

stats = bmark_solvers(solver, ps_problems; max_time, max_eval)
keys_stats = keys(stats)

path_result = pwd()*"/dvpt/comparaison/results/v0_1_0/"


#select fields
selected_fields = [:name, :nvar, :elapsed_time, :iter, :dual_feas, :status, :objective, :neval_obj, :neval_grad, :neval_hprod, :obj_5grad_5Hv, :time_sur_obj_5grad_5Hv, :inverse_pourcentage_pas_accepte]
for i in keys_stats
  println(stdout, "\n" * string(i) )
  pretty_stats(stdout, stats[i][!, [:name, :nvar, :elapsed_time, :iter, :dual_feas, :status, :objective, :neval_obj, :neval_grad, :neval_hprod]], tf=tf_markdown)
end


location_md = string(path_result*"tables.md")
io = open(location_md,"w")
for i in keys_stats
  println(io, "\n" * string(i) )
  pretty_stats(io, stats[i][!, [:name, :nvar, :elapsed_time, :iter, :dual_feas, :status, :objective, :neval_obj, :neval_grad, :neval_hprod]], tf=tf_markdown)
end
close(io)

cost(df) = (df.status .!= :success) * Inf + df.t

# write Dataframes
for key in keys_stats
  CSV.write(path_result*"dataframes/"*string(key)*".csv", stats[key])
end 

# read Dataframes
# memorize_keys_stats = [:PLSR1, :PBFGS, :PSE, :PLSE, :PLBFGS, :TRUNK, :PSR1, :PCS]

# new_stats = Dict{Symbol, DataFrame}()
# for key in memorize_keys_stats
#   new_stats[key] = DataFrame(CSV.File(path_result*"dataframes/"*string(key)*".csv"))
# end 

#= write latex=#
println("écriture des résultats latex")
location_latex = string(path_result*"tables.tex")
io = open(location_latex,"w")
for i in keys_stats
  println(io, "\n" * string(i) )
  pretty_latex_stats(io, stats[i][!, [:name, :nvar, :elapsed_time, :iter, :dual_feas, :status, :objective, :neval_obj, :neval_grad, :neval_hprod]])
end
close(io)

#no profil print, mainly for servers
ENV["GKSwstype"]=100

p_iter = SolverBenchmark.performance_profile(stats, df -> (df.status .!= :first_order) * Inf + df.iter; legend=:bottomright )
savefig(p_iter, path_result*"profiles/iter_profile.pdf")
p_time = SolverBenchmark.performance_profile(stats, df -> (df.status .!= :first_order) * Inf + df.elapsed_time; legend=:bottomright )
savefig(p_time, path_result*"profiles/time_profile.pdf")
```

The script store every `Dataframe` in a CSV file, and you can retrieve them by uncommenting:
```julia
#Lecture
# memorize_keys_stats = [:PLSR1, :PBFGS, :PSE, :PLSE, :PLBFGS, :TRUNK, :PSR1, :PCS]

# new_stats = Dict{Symbol, DataFrame}()
# for key in memorize_keys_stats
#   new_stats[key] = DataFrame(CSV.File(path_result*"dataframes/"*string(key)*".csv"))
# end 
```

For `n=10` the first results are 
[iter_profile.pdf](https://github.com/JuliaSmoothOptimizers/PartiallySeparableSolvers.jl/files/9154549/iter_profile.pdf)
[time_profile.pdf](https://github.com/JuliaSmoothOptimizers/PartiallySeparableSolvers.jl/files/9154550/time_profile.pdf)

On very small problmes, trunk performs better.
What is strange is that partitioned methods are faster, but require more iterations before convergence.
I'm waiting a frontal22 access before change `n` to `1000` and run again to get reprentative profiles.

**Comment by paraynaud** [comment]
@dpo, here are the result when the problems have 1000 variables:
[time_profile.pdf](https://github.com/JuliaSmoothOptimizers/PartiallySeparableSolvers.jl/files/9176247/time_profile.pdf)
[iter_profile.pdf](https://github.com/JuliaSmoothOptimizers/PartiallySeparableSolvers.jl/files/9176248/iter_profile.pdf)

They stay similar.
I am still surprised that partitioned methods are faster but make more iterations than trunk.
On these results, PCS seem to be the best methods among the partitioned methods.

**Reply by dpo** [reply]
Thank you. That is very interesting! Are you running Trunk with the same stopping conditions as the PS solvers? You could use the callback feature to make sure they stop in the same conditions.

Also, there's an orange-brown curve above PCS. Is it PLBFGS?

Could you also extract statistics on the number of elements in each problem, the average number of elemental variables, convexity, etc.?

**Reply by dpo** [reply]
I also meant to add that Trunk has several options, like the non-monotone feature, that should be deactivated (I'm assuming the PS solvers are monotonic).

**Reply by paraynaud** [reply]
> Also, there's an orange-brown curve above PCS. Is it PLBFGS?

It is PSR1.

>  Are you running Trunk with the same stopping conditions as the PS solvers? You could use the callback feature to make sure they stop in the same conditions.

I specified a `max_time` and `max_iter` criterias, but I kept the default atol and rtol.

> Could you also extract statistics on the number of elements in each problem, the average number of elemental variables, convexity, etc.?

I could extract it for each problem similarly to `show` of PartiallySeparableNLPModel and make the stats separately.

> I also meant to add that Trunk has several options, like the non-monotone feature, that should be deactivated (I'm assuming the PS solvers are monotonic).

Should I put `monotone::Bool=false`?

**Reply by dpo** [reply]
`monotone = true`.

Another detail is that we don't see the memory usage. PSR1 and PLSR1 can be quite different.

**Reply by paraynaud** [reply]
Are memory allocations recorded by `NLPModels`?

**Reply by dpo** [reply]
No.

**Comment by paraynaud** [comment]
The last results:
[time_profile.pdf](https://github.com/JuliaSmoothOptimizers/PartiallySeparableSolvers.jl/files/9221244/time_profile.pdf)
[iter_profile.pdf](https://github.com/JuliaSmoothOptimizers/PartiallySeparableSolvers.jl/files/9221243/iter_profile.pdf)

**Reply by paraynaud** [reply]
`trunk` uses exact second derivatives.

**Reply by dpo** [reply]
You can pass an LBFGSModel or LSR1Model to it.

**Comment by paraynaud** [comment]
I compared the partitoned solvers to `trunk` using exacts derivatives or LBFGSNLPModel, here are the results;
[time_profile.pdf](https://github.com/JuliaSmoothOptimizers/PartiallySeparableSolvers.jl/files/9242765/time_profile.pdf)
[iter_profile.pdf](https://github.com/JuliaSmoothOptimizers/PartiallySeparableSolvers.jl/files/9242766/iter_profile.pdf)

Except PBFGS an PLBFGS the partitioned methods have better performances (time and iteration) than LBFGS.
I will add a LSR1NLPModel the next time I launch the comparison.

**Reply by dpo** [reply]
Thanks! So if the P methods need more iterations, they also need more objective/gradient evaluations. Why is that? Are there many rejected steps?

**Reply by paraynaud** [reply]
> So if the P methods need more iterations, they also need more objective/gradient evaluations. Why is that? Are there many rejected steps?

You mean PBFGS and PLBFGS?

I'll think of a way to compare the ratio $\frac{rejected steps}{total steps}$.
`neval_obj - neval_grad` should give the number of rejected steps, maybe a mean of rejected steps by solver is enough for now.

**Reply by dpo** [reply]
For starters, could you show the output of TRUNK/LBFGS and PBFGS on one or two problems?
